### DISCUSSION TITLE
Redis - How kombu actually deals with the `visibility_timeout`

### DISCUSSION THREAD

**Discussion by V0lantis** [celery/kombu#1591] [Q&A] [answered]
Hello,

I am using celery as a queue broker with redis as a PubSub interface. I have an issue with workers shutting down, and I was trying to mitigate this issue. 
While reading the code, I understood that celery was using Kombu under the hood to communicate with Redis. Unfortunately, I haven't been able to understand the code properly, and how Celery + kombu actually implements the `visibility_timeout` process. 
I am not very accustomed on how redis works, so it is even harder to debug what's happening. Though, I have been able to see that Kombu or celery was creating 3 keys when we launch the program : 

```
1) "_kombu.binding.celery"
2) "_kombu.binding.celeryev"
3) "_kombu.binding.celery.pidbox"
```
Are those keys have anything to do with `visibility_timeout`? 
Since, `visibility_timeout` is not natively implemented in redis, how do you create this behavior? 

Thank-you for any help you could give me, or any hint to a blog article explaining this

**Comment by TayfurYldz** [comment, marked as answer]
Those three `_kombu.binding.*` keys are routing/binding metadata; they are not the Redis visibility-timeout mechanism.

Kombu implements visibility timeout for Redis by emulating acknowledgements. In the current Redis transport, when a message is delivered its delivery tag and serialized message are recorded in two structures:

- `unacked` — a hash containing the unacknowledged message data.
- `unacked_index` — a sorted set whose score is the delivery timestamp.

On acknowledgement, Kombu removes the delivery tag from both. Periodically, `restore_visible()` computes:

```python
ceil = time() - self.visibility_timeout
```

and finds old entries in `unacked_index`; `restore_by_tag()` removes the tracking entry and restores the serialized message to the queue.

The current defaults in the Redis channel are `visibility_timeout = 3600`, with configurable restore interval/throttling as well.

So conceptually Redis itself is not hiding a queue item for N seconds. Kombu moves/tracks the in-flight state separately and later restores messages that remained unacknowledged beyond the timeout.

This also means `visibility_timeout` is relevant to a worker that dies after receiving a message but before acknowledging it. The `_kombu.binding.celery`, `celeryev`, and `celery.pidbox` keys you're seeing serve a different purpose: exchange/queue routing information.
